### PR TITLE
Deglobalize LSA

### PR DESCRIPTION
**Pull request by denesb**
This series gets rid of two globals:
* default_dirty_memory_manager
* logalloc shard tracker

The series was motivated by the desire to be able to configure LSA memory on application startup, with parameters that influence how it instantiates and configures certain member objects. This is borderline impossible currently as the LSA state is scattered across a handful of global variables with static initialization, hence their constructors cannot be customized. This series consolidates all the LSA state into a single object: `logalloc::tracker` (except the migrator functions, those are still in a separate global), then makes this a regular sharded service instantiated by main, and propagated to all object that require it (`logalloc::region` objects in memtable and cache). In the process all global objects (except the migrators) are removed.
The need to remove the `default_dirty_memory_manager` (and the `tests_mgr` in `memtable.cc`) arises from the requirement to pass a `logalloc::tracker` as a constructor parameter to `dirty_memory_manager` instances.

Refs: https://github.com/scylladb/scylla/issues/9882

### CODE REVIEW

**Comment by denesb** [comment]
All tests are passing, ready for review.

**Comment by denesb** [comment]
“There can only be one” is true for almost all the globals we are know getting rid of with hard work, regretting making them a global in the first place.

I want to be able to control the segment store created and used by the segment pool. This is not possible with the current static initialization. This is needed by the tools who will be using standard allocator and hence can’t use the regular segment store (it crashes with standard allocator).

**Comment by denesb** [comment]
(rebased and force pushed to trigger CI)

**Comment by avikivity** [comment]
> “There can only be one” is true for almost all the globals we are now getting rid of with hard work, regretting making them a global in the first place. 

This one is inherently so, because it integrates with malloc(). There can only be one malloc(). Two lsas will have to fight over the single integration point.

> I want to be able to control the segment store created and used by the segment pool. This is not possible with the current static initialization. This is needed by the tools who will be using standard allocator and hence can’t use the regular segment store (it crashes with standard allocator).

IIRC only memtable/cache interact with lsa. Do the tools use memtable/cache? They really shouldn't.

**Comment by denesb** [comment]
> On 21 Jun 2022, at 19:39, Avi Kivity ***@***.***> wrote:
> 
> ﻿
> “There can only be one” is true for almost all the globals we are now getting rid of with hard work, regretting making them a global in the first place.
> 
> This one is inherently so, because it integrates with malloc(). There can only be one malloc(). Two lsas will have to fight over the single integration point.
> 

Lsa allocations happen over allocator instances which are polimorphic. Lsa doesn’t override any of the usual memory allocation APIs. Reclaim happens through the database object. Non globals can and do hook into this like the reader concurrency semaphores.
> I want to be able to control the segment store created and used by the segment pool. This is not possible with the current static initialization. This is needed by the tools who will be using standard allocator and hence can’t use the regular segment store (it crashes with standard allocator).
> 
> IIRC only memtable/cache interact with lsa. Do the tools use memtable/cache? They really shouldn't.
> 

Cache and memtable is easy to avoid in the tools but lsa ha trickled down into the sstable layer and got pretty ingrained there. We recently doubled down on this with page cache  and lru for index. Making it opt-out wold be a major refactor and a source of major risk as any mistake could result in alloc-free mismatch between the regular allocator and lsa. Deglobalozing lsa and thus making it actually configurable seemed like the least risky option. This was actually the last thing I tried. I spent a lot of time exploring the two other options:
* make lsa opt-in in the sstable layer (mentioned above)
* make region_impl a proper interface and provide an alternative region implementation which works with standard allocator. This ironically also requires non-global lsa.

**Comment by avikivity** [comment]
```c++
            _cached_index_file = seastar::make_shared<cached_file>(_index_file,
                                                                   index_page_cache_metrics,
                                                                   _manager.get_cache_tracker().get_lru(),
                                                                   _manager.get_cache_tracker().region(),
                                                                   _index_file_size);
            _index_file = make_cached_seastar_file(*_cached_index_file);
```

This one is easily avoided by just using a regular file. But the object caches are harder to avoid.

**Comment by denesb** [comment]
> Ah, I forgot about those caches. @tgrabiec any opinion about whether it can be disentangled? I agree the simplest route is probably keeping the use of lsa (but perhaps it's enough to just initialize it in main()).

We already have initialization in main, in the form of `logalloc::tracker::init()` and `prime_segment_pools()`. We could in theory expand this to be also used by tools to configure the segment pool used. We could probably get away with this as there should be no segment used before we call these methods. We'd still need my series though, minus the last two patches. The question then becomes: given that I already did the work, is it worth keeping LSA global? I admit that right now I don't expect a lot of benefits from a deglobalized LSA, beyond a saner initialization and teardown (which we currently don't do at all). The tests *did* benefit from the deglobalization: in particular logalloc test can now be run test-case by test case as we now have clear init-teardown sequences in each tests, we don't rely on running the tests in a particular order. I can also see this deglobalization bring us closer to the scylla-in-a-box goal.
I don't see much in terms of cost of deglobalization: it is mainly inconvenient for tests, who now has to create and pass instances of `logalloc::tracker` or `dirty_memory_manager`. Impact on production code is surprisingly limited because we already propagate region instances everywhere LSA is used and these instances in turn are all created in `database` which has access to the tracker instance.
All this is to say: I don't insist on pushing this in, all I really want is to achieve my goal of controlling the segment pool creation, which can be achieved with the first N-2 patches of my series. I think overall the deglobalization is beneficial for us, and the work is already done.

**Comment by avikivity** [comment]
lsa segment pools are inherently one-per-shard. So it keep being global - you can't construct two. It still brings benefit in that you can't use it before you initialize it (both annoying for tests since you have to add boilerplate, and also good for tests because you can't forget to add initialization boilerplate, it won't compile).

**Comment by denesb** [comment]
(rebased on master)

**Comment by avikivity** [comment]
> > lsa segment pools are inherently one-per-shard. So it keep being global - you can't construct two. It still brings benefit in that you can't use it before you initialize it (both annoying for tests since you have to add boilerplate, and also good for tests because you can't forget to add initialization boilerplate, it won't compile).
> 
> Yes. We have many such pseudo-globals: all our `sharded<>` in main are such: you won't ever have two database objects per shard but it being a local variable in main allows us to have a sane startup/shutdown sequence, as well as pass configuration. Exactly the things this series achieves.

It's only partially the same. We can have multiple sharded<database> if we emulate several nodes in one process. We can't have multiple shard segment pools in one shard no matter how much we want to. The dependency management part is the same.

**Comment by denesb** [comment]
Rebased and force pushed again with the hope of finally getting a CI run that finishes.

**Comment by denesb** [comment]
> It's only partially the same. We can have multiple sharded if we emulate several nodes in one process. We can't have multiple shard segment pools in one shard no matter how much we want to. The dependency management part is the same.

So... where does that leave us? I ran out of arguments and I have no idea whether I convinced you or not. Will you merge this as-is or should I drop the deglobalization bits? I stand by my opinion that deglobalizing LSA is beneficial overall, even if it still remains a global after the series, just one with a defined life-cycle.

**Comment by denesb** [comment]
Debug tests have problems, I will have to investigate.

**Comment by denesb** [comment]
@avikivity I will split this in two PRs: consolidating the state into a single object (`logalloc::tracker`) and the deglobalization. The former is less controversial and getting it in will help with the rebase-treadmill.

File: api/storage_service.cc [hunk 1]
```diff
@@ -1390,25 +1390,23 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
             });
         }
 
-        sstables::compaction_type_options::scrub opts = {
-            .operation_mode = scrub_mode,
-        };
         const sstring quarantine_mode_str = req_param<sstring>(*req, "quarantine_mode", "INCLUDE");
+        sstables::compaction_type_options::scrub::quarantine_mode quarantine_mode;
         if (quarantine_mode_str == "INCLUDE") {
```
Comment (avikivity) [inline api/storage_service.cc:1395]: This patch seems unrelated.

File: compaction/compaction_manager.cc [hunk 2]
```diff
@@ -1183,7 +1183,7 @@ class compaction_manager::validate_sstables_compaction_task : public compaction_
                     sst->get_sstable_level(),
                     sstables::compaction_descriptor::default_max_sstable_bytes,
                     sst->run_identifier(),
-                    sstables::compaction_type_options::make_scrub(sstables::compaction_type_options::scrub::mode::validate));
+                    sstables::compaction_type_options::make_scrub(sstables::compaction_type_options::scrub::mode::validate, _compacting_table->logalloc_tracker()));
```
Comment (avikivity) [inline compaction/compaction_manager.cc:1186]: Ho hum. Instead of passing logalloc::tracker and make everyone scratch their head, how about passing a function that makes memtables? That will push the odd dependency to the caller.

We should have hidden the memtable completely in some way, but that's harder and this is really an ad-hoc use case.

File: db/virtual_table.hh [hunk 3]
```diff
@@ -53,7 +53,7 @@ public:
 class memtable_filling_virtual_table : public virtual_table {
     dirty_memory_manager _dmm;
 public:
-    using virtual_table::virtual_table;
+    explicit memtable_filling_virtual_table(schema_ptr s, logalloc::tracker& tracker) : virtual_table(std::move(s)), _dmm(tracker) { }
```
Comment (avikivity) [inline db/virtual_table.hh:56]: Similarly here. Pass a memtable-creating function instead of logalloc::tracker which is too low-level here.

File: db/system_keyspace.cc [hunk 4]
```diff
@@ -2332,8 +2332,8 @@ class runtime_info_table : public memtable_filling_virtual_table {
 
 class versions_table : public memtable_filling_virtual_table {
 public:
-    explicit versions_table()
-        : memtable_filling_virtual_table(build_schema()) {
+    explicit versions_table(replica::database& db)
```
Comment (avikivity) [inline db/system_keyspace.cc:2335]: We should avoid passing replica::database which is replica-side to a coordinator-side class.

Comment (denesb) [reply db/system_keyspace.cc:2335]: Virtual tables are replica side. These are later wrapped by `replica::table` objects.

File: api/storage_service.cc (hunk 1 above)
Comment (denesb) [reply api/storage_service.cc:1395]: It prepares the ground for passing passing lower-level arguments to scrub compaction. Currently the options struct mixes high level and low level options, which I have to separate before doing that.

File: db/system_keyspace.cc (hunk 4 above)
Comment (avikivity) [reply db/system_keyspace.cc:2335]: Ah, ok.

**Comment by denesb** [comment]
Fixed another debug mode compilation failure that somehow slipped through.

**Comment by denesb** [comment]
I'm closing this. This is too much to get in at once, I struggle to keep up with rebasing. Will try to get in smaller chunks.

**Comment by avikivity** [comment]
I think you can avoid the parts moved to dirty_memory_manager. I'm trying to disappear them.

**Comment by denesb** [comment]
What are you trying to get rid of? The region group (currently in dirty_memory_manager.hh).
Yes, I think I'll freeze this effort for now. I think I can get what I needed for the tools with just https://github.com/scylladb/scylladb/pull/11284 (hidden review ping). My only worry is that once tools have a workaround I'll lack the motivation to resurrect this later. But we'll see.

**Comment by avikivity** [comment]
region_group will lost most of each functionality and the remainder will be absorbed into dirty_memory_manager.
